### PR TITLE
UCT/RC/DC: Allocate zero-copy send ops from a freelist.

### LINKED ISSUE #1440: Segmentation fault with rc on the many_isend test
There is a seg fault that happens when running on a very large scale with ucx. with np=320, for example it didn't reproduce. In mtt, it reproduced with np=2432.
Happends on every run with with rc_x and with dc_x. Passes with ud and ud_x.

The command line to reproduce:

mpirun -np 2432 -mca btl_openib_warn_default_gid_prefix 0 --bind-to core --tag-output --timestamp-output --display-map -mca pml ucx -x UCX_NET_DEVICES=mlx5_0:1 -mca btl_openib_if_include mlx5_0:1 -mca coll_hcoll_enable 0 -x UCX_TLS=rc_x -mca opal_pmix_base_async_modex 1 -mca mpi_add_procs_cutoff 0 -mca pmix_base_collect_data 0 --map-by node /hpc/scrap/users/mtt/scratch/ucx_ompi/20170418_221848_20109_725574_clx-hercules-001/installs/KGVi/tests/mpich_tests/mpich-mellanox.git/test/mpi/pt2pt/many_isend

The trace is always the same:
```
Wed Apr 19 09:53:10 2017[1,722]<stderr>:==== backtrace ====
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 2 0x0000000000067b1c mxm_handle_error()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/mxm-master/src/mxm/util/debug/debug.c:641
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 3 0x000000000006806c mxm_error_signal_handler()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/mxm-master/src/mxm/util/debug/debug.c:616
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 4 0x0000000000035670 killpg()  ??:0
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 5 0x000000000002bbdf uct_rc_txqp_completion_desc()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/rc/base/rc_ep.h:318
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 6 0x000000000002bbdf uct_rc_mlx5_txqp_process_tx_cqe()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/rc/accel/rc_mlx5_common.h:96
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 7 0x000000000002bbdf uct_rc_mlx5_iface_poll_tx()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/rc/accel/rc_mlx5_iface.c:62
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 8 0x000000000002bbdf uct_rc_mlx5_iface_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/ib/rc/accel/rc_mlx5_iface.c:75
Wed Apr 19 09:53:10 2017[1,722]<stderr>: 9 0x000000000001917e ucs_callbackq_dispatch()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucs/datastruct/callbackq.inl:39
Wed Apr 19 09:53:10 2017[1,722]<stderr>:10 0x000000000001917e uct_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/uct/base/uct_md.c:233
Wed Apr 19 09:53:10 2017[1,722]<stderr>:11 0x000000000000f11d ucp_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucp/core/ucp_worker.c:719
Wed Apr 19 09:53:10 2017[1,722]<stderr>:12 0x000000000000f11d ucs_async_check_miss()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucs/async/async.h:75
Wed Apr 19 09:53:10 2017[1,722]<stderr>:13 0x000000000000f11d ucp_worker_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ucx-master/src/ucp/core/ucp_worker.c:720
Wed Apr 19 09:53:10 2017[1,722]<stderr>:14 0x0000000000002c77 mca_pml_ucx_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mca/pml/ucx/pml_ucx.c:421
Wed Apr 19 09:53:10 2017[1,722]<stderr>:15 0x000000000002eecc opal_progress()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/opal/runtime/opal_progress.c:225
Wed Apr 19 09:53:10 2017[1,722]<stderr>:16 0x0000000000002fbd mca_pml_ucx_recv()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mca/pml/ucx/pml_ucx.c:510
Wed Apr 19 09:53:10 2017[1,722]<stderr>:17 0x000000000005b354 PMPI_Recv()  /hpc/local/benchmarks/hpcx_install_Tuesday/src/hpcx-gcc-redhat7.2/ompi-v2.x/ompi/mpi/c/profile/precv.c:77
Wed Apr 19 09:53:10 2017[1,722]<stderr>:18 0x00000000004025db main()  ??:0
Wed Apr 19 09:53:10 2017[1,722]<stderr>:19 0x0000000000021b15 __libc_start_main()  ??:0
Wed Apr 19 09:53:10 2017[1,722]<stderr>:20 0x00000000004023e9 _start()  ??:0
Wed Apr 19 09:53:10 2017[1,722]<stderr>:===================
Wed Apr 19 09:53:10 2017[1,1891]<stderr>:[clx-hercules-098:30965:0] Caught signal 11 (Segmentation fault)
```

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #1440 and #1266

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1530/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1531/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3451/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3452/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1532/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3453/ for details (Mellanox internal link).

File: src/uct/ib/rc/base/rc_iface.h [hunk 1]
```diff
@@ -215,6 +225,7 @@ struct uct_rc_iface_send_op {
     ucs_queue_elem_t              queue;
     uct_rc_send_handler_t         handler;
     uint16_t                      sn;
+    uint16_t                      flags;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: right

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1535/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3456/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1536/ for details.

**Pull request merged**